src/quickcheck.rs

```diff
@@ -0,0 +1,264 @@
+        let shrunk_minute = orig_m
+            .shrink()
+            .map(move |m| Self::from_hms_nanos_unchecked(orig_h, m, orig_s, orig_nanos));
+        let shrunk_second = orig_s
+            .shrink()
+            .map(move |s| Self::from_hms_nanos_unchecked(orig_h, orig_m, s, orig_nanos));
+        let shrunk_nanos = orig_nanos
+            .shrink()
+            .map(move |nanos| Self::from_hms_nanos_unchecked(orig_h, orig_m, orig_s, nanos));
+
+        Box::new(
+            shrunk_hour
+                .chain(shrunk_minute)
+                .chain(shrunk_second)
+                .chain(shrunk_nanos),
+        )
+    }
+}
+
+impl Arbitrary for PrimitiveDateTime {
+    fn arbitrary<G>(g: &mut G) -> Self
+    where
+        G: Gen,
+    {
+        let d = Date::arbitrary(g);
+        let t = Time::arbitrary(g);
+        Self::new(d, t)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        let orig_d = self.date;
+        let orig_t = self.time;
+
+        let shrunk_date = orig_d.shrink().map(move |d| Self::new(d, orig_t));
+        let shrunk_time = orig_t.shrink().map(move |t| Self::new(orig_d, t));
+
+        Box::new(shrunk_date.chain(shrunk_time))
+    }
+}
+
+impl Arbitrary for UtcOffset {
+    fn arbitrary<G>(g: &mut G) -> Self
+    where
+        G: Gen,
+    {
+        let size = min(
+            i32::try_from(g.size()).unwrap_or(i32::max_value()),
+            60 * 60 * 24,
+        );
+        let o = g.gen_range(-(size - 1), size);
+        Self::seconds_unchecked(o)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        Box::new(self.as_seconds().shrink().map(Self::seconds_unchecked))
+    }
+}
+
+impl Arbitrary for OffsetDateTime {
+    fn arbitrary<G>(g: &mut G) -> Self
+    where
+        G: Gen,
+    {
+        let dt = PrimitiveDateTime::arbitrary(g);
+        let o = UtcOffset::arbitrary(g);
+        dt.assume_offset(o)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        let orig_dt = PrimitiveDateTime::new(self.date(), self.time());
+        let orig_o = self.offset();
+
+        let shrunk_datetime = orig_dt.shrink().map(move |dt| dt.assume_offset(orig_o));
+        let shrunk_offset = orig_o.shrink().map(move |o| orig_dt.assume_offset(o));
+
+        Box::new(shrunk_datetime.chain(shrunk_offset))
+    }
+}
+
+impl Arbitrary for Weekday {
+    fn arbitrary<G>(g: &mut G) -> Self
+    where
+        G: Gen,
+    {
+        use Weekday::*;
+        match g.gen_range(0, max(1, min(7, g.size()))) {
+            0 => Monday,
+            1 => Tuesday,
+            2 => Wednesday,
+            3 => Thursday,
+            4 => Friday,
+            5 => Saturday,
+            _ => Sunday,
+        }
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        match self {
+            Weekday::Monday => Box::new(std::iter::empty()),
+            _ => Box::new(std::iter::once(self.previous())),
```

Comment:
Having not tested this, is `shrink` called multiple times? From looking at the other types, it looks like a complete iterator is returned on the first call, where only a single item is returned here.

Reply:
Yes, quickcheck uses `shrink` to iteratively minimize the test case, and it calls the method multiple times while doing so. I'm honestly not sure whether it's better to return "long" iterators at a time or just one, or how much it matters. I considered using `iter::successors` here but ultimately decided against it for no particular reason.

Reply:
If performance somehow becomes an issue with this, then it can be handled at that point. As long as the behavior is correct, it doesn't matter currently.